test/src/apps/devfs/framebuffer.c

```diff
@@ -0,0 +1,214 @@
+
+	memset(red, 0, sizeof(red));
+	memset(green, 0, sizeof(green));
+	memset(blue, 0, sizeof(blue));
+
+	struct fb_cmap get_cmap = {
+		.start = 0,
+		.len = cmap_len,
+		.red = red,
+		.green = green,
+		.blue = blue,
+		.transp = NULL,
+	};
+
+	TEST_SUCC(ioctl(fb_fd, FBIOGETCMAP, &get_cmap));
+	TEST_RES(memcmp(red, red_expected, sizeof(red)), _ret == 0);
+	TEST_RES(memcmp(green, green_expected, sizeof(green)), _ret == 0);
+	TEST_RES(memcmp(blue, blue_expected, sizeof(blue)), _ret == 0);
+
+	// Test invalid color map: use a start index that's definitely out of range
+	// Most devices have 256 entries or less, so start=0x10000 should always fail
+	struct fb_cmap invalid_cmap = {
+		.start = 0x10000,
+		.len = 1,
+		.red = red,
+		.green = green,
+		.blue = blue,
+		.transp = NULL,
+	};
+
+	TEST_ERRNO(ioctl(fb_fd, FBIOPUTCMAP, &invalid_cmap), EINVAL);
+}
+END_TEST()
+
+FN_TEST(write_enospc)
+{
+	const unsigned char test_pattern = 0xff;
+
+	if (fb_smem_len == 0) {
+		fprintf(stderr,
+			"write_enospc: framebuffer has zero length; skipping test\n");
+		return;
+	}
+
+	// First, read the last byte to preserve it
+	unsigned char original_byte = 0;
+	TEST_RES(lseek(fb_fd, (off_t)(fb_smem_len - 1), SEEK_SET),
+		 _ret == (off_t)(fb_smem_len - 1));
+	TEST_RES(read(fb_fd, &original_byte, 1), _ret == 1);
+
+	// Now seek to the end and try to write beyond the end of the framebuffer
+	TEST_RES(lseek(fb_fd, (off_t)fb_smem_len, SEEK_SET),
+		 _ret == (off_t)fb_smem_len);
+	TEST_ERRNO(write(fb_fd, &test_pattern, sizeof(test_pattern)), ENOSPC);
+
+	// Finally, verify that the last byte is unchanged
+	TEST_RES(lseek(fb_fd, (off_t)(fb_smem_len - 1), SEEK_SET),
+		 _ret == (off_t)(fb_smem_len - 1));
+	unsigned char check_byte = 0;
+	TEST_RES(read(fb_fd, &check_byte, 1), _ret == 1);
+	TEST_RES(memcmp(&check_byte, &original_byte, sizeof(check_byte)),
+		 _ret == 0);
+}
+END_TEST()
+
+FN_TEST(mmap_mremap_and_fork)
+{
+	static uint8_t pattern[PAGE_SIZE];
+	static uint8_t fork_pattern[PAGE_SIZE];
+
+	size_t map_len = fb_smem_len;
+	if (map_len > sizeof(pattern)) {
+		map_len = sizeof(pattern);
+	}
+
+	if (map_len == 0) {
+		fprintf(stderr,
+			"mmap_mremap_and_fork: framebuffer has zero length; skipping test\n");
+		return;
+	}
+
+	for (size_t i = 0; i < map_len; ++i) {
+		pattern[i] = (uint8_t)(i & 0xff);
+		fork_pattern[i] = (uint8_t)(0xaa ^ (i & 0xff));
+	}
+
+	uint8_t *mapped =
+		TEST_RES((uint8_t *)mmap(NULL, map_len, PROT_READ | PROT_WRITE,
+					 MAP_SHARED, fb_fd, 0),
+			 _ret != MAP_FAILED);
+	memcpy(mapped, pattern, map_len);
+	TEST_RES(memcmp(mapped, pattern, map_len), _ret == 0);
+
+	uint8_t *remapped = TEST_RES((uint8_t *)mremap(mapped, map_len, map_len,
+						       MREMAP_MAYMOVE),
+				     _ret != MAP_FAILED);
+	if (remapped != mapped) {
+		mapped = remapped;
+	}
+
+	TEST_RES(memcmp(mapped, pattern, map_len), _ret == 0);
+
+	pid_t pid = CHECK(fork());
+	if (pid == 0) {
+		// Child process: avoid TEST_* so the parent sees the status via waitpid
```

Comment:
Yes, so please use `CHECK`.
```suggestion
CHECK_WITH(memcmp(mapped, pattern, map_len), _ret == 0);
```